rohmu/object_storage/local.py

```diff
@@ -244,6 +251,82 @@ def store_file_object(
             key=key, size=os.path.getsize(target_path), metadata=self.sanitize_metadata(self._filter_metadata(metadata))
         )
 
+    def create_concurrent_upload(
+        self, key: str, metadata: Optional[Metadata] = None, mimetype: Optional[str] = None
+    ) -> ConcurrentUploadId:
+        upload_id = uuid.uuid4().hex
+        chunks_dir = self.format_key_for_backend("concurrent_upload_" + upload_id)
+        os.makedirs(chunks_dir, exist_ok=True)
+        self.stats.operation(StorageOperation.create_multipart_upload)
+
+        upload = ConcurrentUploadData("local", upload_id, key)
+        upload_id = self._create_upload_id(upload)
+        self._concurrent_uploads[upload_id] = (upload, metadata, {})
+        return upload_id
+
+    def _get_upload_info(
+        self, upload_id: ConcurrentUploadId
+    ) -> tuple[ConcurrentUploadData, Optional[Metadata], dict[int, str]]:
+        try:
+            return self._concurrent_uploads[upload_id]
+        except KeyError:
+            upload = self._parse_upload_id(upload_id)
+            return upload, None, {}
+
+    def upload_concurrent_chunk(
+        self,
+        upload_id: ConcurrentUploadId,
+        chunk_number: int,
+        fd: BinaryIO,
+        upload_progress_fn: IncrementalProgressCallbackType = None,
+    ) -> None:
+        concurrent_data, _, chunks = self._get_upload_info(upload_id)
+        chunks_dir = self.format_key_for_backend("concurrent_upload_" + concurrent_data.backend_id)
+        try:
+            with atomic_create_file_binary(os.path.join(chunks_dir, str(chunk_number))) as chunk_fp:
+                wrapped_fd = ProgressStream(fd)  # type: ignore[abstract]
```

Comment:
The mypy comment is necessary because it's not smart enough to understand ProgressStream dynamic method definitions....